portfolio/src/main/java/com/google/sps/servlets/CommentsHandler.java

```diff
@@ -88,11 +88,34 @@ public String getComments() {
     return gson.toJson(commentsMap.values());
   }
 
-  public void deleteComments() {
+  // deletes the comment with the given key
+  // deletes all child comments
+  public void deleteComment(String key) {
+    // Technically if there are replies to this comment I can delete the whole comments thread, I don't know
+    // what is better - to delete them all or to change parent comment to dummy one with "deleted" text
+    // Here I am implementing the first one
+    Map<Key, Entity> commentsMap = new HashMap<>();
     Query query = new Query("Comment");
     PreparedQuery results = datastore.prepare(query);
-    for (Entity entity : results.asIterable()) {
-      datastore.delete(entity.getKey());
+    for (Entity entity: results.asIterable()) {
+      commentsMap.put(entity.getKey(), entity);
+    }
+    deleteThread(KeyFactory.stringToKey(key), commentsMap);
+  }
+  
+  // deletes current comment and recursively calls itself to delete all
+  // child comments
+  private void deleteThread(Key current, Map<Key, Entity> commentsMap) {
+    datastore.delete(current);
+    // this can be done faster, presumably in O(thread length) if I store reversed reply keys in the datastore,
```

Comment:
Generally, latency sensitive systems do a lazy delete. They set a field called "deleted" (or whatever) to true and then they don't display it at all. Things marked deleted are slowly deleted over time.  

PS: This is just an FYI. I am not asking you implement anything like that.

Reply:
In real production system that may also be required for auditing purpose to keep the deleted content for some time too.